File: go/primeSieve.go
package main

type primeSieve []bool

func getSieve(sieveSize int) primeSieve {
	sieve := make([]bool, sieveSize+1)
	for i := range sieve {
		sieve[i] = true
	}
	return sieve
}

func (sieve primeSieve) run() {
	for factor := 2; factor < len(sieve); factor++ {
		if sieve[factor] {
			q := factor + factor
			for q < len(sieve) {
				sieve[q] = false
				q += factor
			}
		}
	}
}

func (sieve primeSieve) result() []int {
	current := 2
	var primes []int

	for _, isPrime := range sieve[2:] {
		if isPrime {
			primes = append(primes, current)
		}
		current++
	}

	return primes
}

File: javascript/primeSieve.js
exports.getSieve = (sieveSize) => {
	const sieve = new Array(sieveSize + 1);

	for (let i = 0; i < sieveSize; i++) {
		sieve[i] = true;
	}

	return {
		run: () => {
			for (let factor = 2; factor <= sieveSize; factor++) {
				if (sieve[factor]) {
					let q = factor + factor;
					while (q <= sieveSize) {
						sieve[q] = false;
						q += factor;
					}
				}
			}
		},
		results: () => {
			let current = 2;
			let primes = [];
			sieve.slice(2).forEach((isPrime) => {
				if (isPrime) {
					primes.push(current);
				}
				current++;
			});

			return primes;
		},
	};
};
